src/pages/cart.tsx

```diff
@@ -0,0 +1,237 @@
+    const [list, settList] = useState<product[]>([]);
+    const [maxElems, setMaxElems] = useState<number>(1);
+    const [pageIndex, setPageIndex] = useState<number>(1);
+    const [pagination, setPagination] = useState<number[]>([]);
+    const [products, setproductsLen] = useState<{length:number,total:number}>({length:0,total:0});
+    const [discount, setDiscount] = useState<number>(0);
+    const [promCodes, setPromCodes] = useState<{val: string; list: string[]}>({val: "",list:[]});
+    const [hidden, setHidden] = useState<boolean>(false);
+    const promo = useRef<HTMLDivElement>(null);
+    const promoInput = useRef<HTMLInputElement>(null);
+    const maxInput = useRef<HTMLInputElement>(null);
+
+    const {state} = useLocation();
+    
+    
+    useEffect(():void => {
+        const st = (state === null)?0:state.id
+        if(st > 0)show()
+        getList()
+        setMaxElems(2)
+        let products : product[] = JSON.parse(localStorage.getItem('cart')  + "")
+        settList(products)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [settest])
+
+    function show(){
+        setHidden(true)
+    }
+
+    useEffect(():void => {
+        let i: number = 0;
+        let p: number[] = [];
+        for(i = 0; i < Math.floor(list.length / maxElems); i++){
+            p.push(i+1);
+        }
+        if(Math.floor(list.length / maxElems) !== list.length / maxElems) p.push(i+1);
+        setPagination(p);
+
+        if(p.length  < pageIndex && pagination.length !== 0) {
+            setPageIndex(p.length)
+        }
+    }, [maxElems, list.length])
+
+    useEffect(():void => {
+        maxInput.current!.value = (getParam("max") === 'null')?'2':getParam("max")
+        localStorage.setItem("cart", JSON.stringify(list));
+        props.clc();
+        let products : product[] = JSON.parse(localStorage.getItem('cart')  + "")
+        let len = 0
+        let total = 0
+        products.forEach(e=>{
+            len += e.amount
+            total += e.amount * e.price
+        })
+        if(localStorage.getItem("promos") === null) localStorage.setItem("promos","[]")
+        setproductsLen({length:len,total:total})
+        setDiscount(total - (total/100 * 10 * promCodes.list.length))
+        setPromCodes({val: "", list: JSON.parse(localStorage.getItem("promos")+"")})
+        }, [list])
+
+    function setTo(event: {target : HTMLInputElement}):void{
+        setUrl("max",event.target!.value)
+        setMaxElems((event.target.value === "")?1:parseInt(event.target.value));
+        getList();
+    }
+
+    function updateList():void{
+        let s: product = {
+            id: 0,title: "", description: "", price: 0, discountPercentage:0, rating: 0, stock: 0, brand: "", category: "",thumbnail: "",images: [], amount: 1
+        }
+        list.push(s)
+        settList(list.filter(elem=> elem.id !== 0).sort(function(a : product, b: product){return b.amount - a.amount}))
```

Comment:
You can move this function function(a : product, b: product){return b.amount - a.amount} in constant